Conventional middleware class requirements. (#12895)
aspnet/AspNetCore.Docs: aspnetcore/fundamentals/middleware/write.md

The preceding sample code is used to demonstrate creating a middleware component. For ASP.NET Core's built-in localization support, see <xref:fundamentals/localization>.

> You can test the middleware by passing in the culture. For example, `https://localhost:5001/?culture=no`.

Reviewer: The "you" is optional here, so it can be dropped ...

```suggestion
Test the middleware by passing in the culture. For example, request `https://localhost:5001/?culture=no`.
```

The following code moves the middleware delegate to a class:

[!code-csharp[](index/snapshot/Culture/RequestCultureMiddleware.cs)]

The middleware class must include:

* A public constructor with a parameter of type <xref:Microsoft.AspNetCore.Http.RequestDelegate>.
* A public method named `Invoke` or `InvokeAsync`. This method must:
  * Return a `Task`.
  * Accept a first parameter of type <xref:Microsoft.AspNetCore.Http.HttpContext>.
  

> Any additional constructor and `Invoke`/`InvokeAsync` parameters will be populated by [dependency injection (DI)](xref:fundamentals/dependency-injection).

Reviewer: Use present tense where possible ...

```suggestion
Any additional constructor and `Invoke`/`InvokeAsync` parameters are populated by [dependency injection (DI)](xref:fundamentals/dependency-injection).
```

## Middleware dependencies

Middleware should follow the [Explicit Dependencies Principle](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#explicit-dependencies) by exposing its dependencies in its constructor. Middleware is constructed once per *application lifetime*. See the [Per-request dependencies](#per-request-dependencies) section if you need to share services with middleware within a request.

Middleware components can resolve their dependencies from [dependency injection (DI)](xref:fundamentals/dependency-injection) through constructor parameters. [UseMiddleware&lt;T&gt;](/dotnet/api/microsoft.aspnetcore.builder.usemiddlewareextensions.usemiddleware#Microsoft_AspNetCore_Builder_UseMiddlewareExtensions_UseMiddleware_Microsoft_AspNetCore_Builder_IApplicationBuilder_System_Type_System_Object___) can also accept additional parameters directly.

> ### Per-request dependencies

Reviewer: I think this works better as an H2 or remove this heading and let the content live under *Middleware dependencies*. What's not ideal in outlining is to have a single heading entry at a given level, which is called out in the Chicago and other style manuals.

  * Return a `Task`.
  * Accept a first parameter of type <xref:Microsoft.AspNetCore.Http.HttpContext>.
  

> Additional constructor and `Invoke`/`InvokeAsync` parameters are populated by [dependency injection (DI)](xref:fundamentals/dependency-injection).

Reviewer: pluralize "constructor" ...

```suggestion
Additional constructors and `Invoke`/`InvokeAsync` parameters are populated by [dependency injection (DI)](xref:fundamentals/dependency-injection).
```

Author: @guardrex It's not really meant to be additional constructors. I realise you did suggest this first time through, but I'm trying to state that parameters for the constructor already mentioned and for `Invoke`/`InvokeAsync` are taken from DI. There's no reason to have more than one constructor for the middleware class. Sorry if I'm still not explaining that well enough...

Reviewer: I see. No ... I see what you mean. It was just that when you had "any" I didn't realize that you only meant one.

Yes on that :point_up: change. That looks great.

Author: Ok, great. I'll make that change now.

Thanks again for this prompt service you're providing. :laughing:

Reviewer: Prompt and **_almost accurate_** review services! ... that's what we do. :smile:
